docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,147 @@
+# Metainfo RPC Service Refactor
+
+## Background
+
+Our current metainfo rpc service isn't very future proof.
+If we shipped our current client code, it wouldn't be easy to update clients and add support for the following:
+- versions
+- multipart uploads
+- cleanup for uncommitted segments
+- general concurrency fixes
+
+Native multipart upload can't happen with our current architecture because there's no concept of starting and ending a stream.
+
+Versions also wouldn't work, because if concurrent clients tried to upload at the same time, they would clobber each other.
```

Comment:
It's the other way around, concurrent uploading causes segments to get mangled and versions is a good way to handle concurrent uploads.